app/frontend/javascript/vue_app.js

```diff
@@ -0,0 +1,196 @@
+import QcInformation from '@/javascript/qc-information/components/QcInformation.vue'
+import TubesToRack from '@/javascript/tubes-to-rack/components/TubesToRack.vue'
+import ValidatePairedTubes from '@/javascript/validate-paired-tubes/components/ValidatePairedTubes.vue'
+
+import MainContent from '@/javascript/shared/components/MainContent.vue'
+import Page from '@/javascript/shared/components/Page.vue'
+import Sidebar from '@/javascript/shared/components/Sidebar.vue'
+
+Vue.use(BootstrapVue)
+Vue.use(BootstrapVueIcons)
+Vue.component('LbMainContent', MainContent)
+Vue.component('LbPage', Page)
+Vue.component('LbSidebar', Sidebar)
+
+export const missingUserIdError = `
+    Unfortunately Limber can't find your user id, which is required to add custom metadata.
+    Click log out and swipe in again to resolve this.
+    If this problem occurs repeatedly, let us know.`
+
+/**
+ * Helper function to initialize and render a Vue component.
+ *
+ * @param {string} selector - The CSS selector of the DOM element to mount the Vue instance on.
+ * @param {Object} component - The Vue component to render.
+ * @param {Object} [props={}] - The props to pass to the Vue component.
+ * @param {Object} [data={}] - The data to pass to the Vue instance.
+ * @param {boolean} [userIdRequired=false] - Whether a user ID is required to render the component.
+ * @returns {Vue} The Vue instance.
+ */
+
+export const renderVueComponent = (selector, component, props = {}, userIdRequired = false, setData = () => ({})) => {
+  const selector_val = `#${selector}`
+  const element = document.querySelector(selector_val)
+  const userId = cookieJar(document.cookie).user_id
+  if (!element) {
+    console.error(`Element with selector ${selector_val} not found.`)
+    return
+  }
+
+  let app
+  if (userIdRequired && !userId) {
+    console.error('User id is required to render this component.')
+    app = new Vue({
+      el: selector_val,
+      render: (h) => h('div', missingUserIdError),
+    })
+  } else {
+    props.userId = userId
+    const data = setData(props)
+    app = new Vue({
+      el: selector_val,
+      data: () => data,
+      render: (h) => h(component, { props }),
+    })
+  }
+  return app
+}
+
+/**
+ * List of elements to initialize as Vue components.
+ * Each element should have an id, a Vue component, an optional flag indicating whether a user ID is required and a setData method.
+ */
+const elements = [
+  {
+    id: 'asset-comments',
+    component: AssetComments,
+    userIdRequired: true,
+    setData: (props) => initialiseCommentFactory(props),
+  },
+  {
+    id: 'asset-comments-counter',
+    component: AssetCommentsCounter,
+    userIdRequired: true,
+    setData: (props) => initialiseCommentFactory(props),
+  },
+  {
+    id: 'asset-comments-add-form',
+    component: AssetCommentsAddForm,
+    userIdRequired: true,
+    setData: (props) => initialiseCommentFactory(props),
```

Comment:
This is a nice refactor but it is still a bit of a code smell. I think the components should be able to function in isolation without manually overriding their data. Maybe the comment factory should be initialised in each of the components in a setup? I don't see a reason to override the data unless there is multiple types of data being passed in.

I know that may be more work so happy to leave it here if you think this story is growing too much.

Reply:
I have actually started working in that way, where the commentFactory is created within components on mounting by passing the required fields fir its creation as props. However, since I don’t fully know the whole design, I was a bit concerned that this might cause unforeseen repercussions. So, I thought of keeping it this way to avoid changing the data part. But from what I’ve seen, all these components work independently, and theoretically, it should be fine for commentFactory to reside within each component.
Do you think it is good to put it down as a separate story?

Reply:
Yeah that makes sense, looks fine to leave as is. Not sure if it is worth its own story.